### PR TITLE
Transactor: Journal client interface

### PR DESCRIPTION
**Pull request by vyzo**
Fleshes out the interface for Journal clients and event listeners and refactors the tests to use it.

Specifics:
- futurize Journal interface methods
- add JournalClient < Journal interface for client objects
- rename JournalClient to JournalListener
- add Copycat.Client class implementing JournalClient
- refactor tests to use the JournalClient interface

### CODE REVIEW

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 1]
```diff
@@ -38,15 +40,61 @@ object Copycat {
     }
   }
   
+  class Client(client: CopycatClient) extends JournalClient {
+    private var listeners: Set[JournalListener] = Set()
+    
+    // Journal 
+    def insert(rec: CanonicalRecord) =
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: can you annotate these with the return types?

Comment (parkan) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: ah I guess I can just look at the trait, that's OK

Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: seconding type annotations. help to have it at the site

Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: ok, will add.

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 2]
```diff
@@ -38,15 +40,61 @@ object Copycat {
...
+    def connect(address: String) {
+      client.connect(new Address(address)).join()
+    }
+    
+    def close() {
+      client.close().join()
+    }
+    
+    def listen(listener: JournalListener) {
+      if (listeners.isEmpty) {
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:72]: this is single threaded, yeah? or are there multiple event loops in a thread pool type situation

Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:72]: yeah, made the client calls synchronous.
Also, there is a problem with the FutureConverters -- the return CompletbleFuture<Void> which is not convertible to Future[Unit].

File: transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala [hunk 3]
```diff
@@ -35,26 +37,30 @@ object JournalBlockchainSpec extends io.mediachain.BaseSpec
...
     val context = JournalBlockchainSpecContext.context
-    val res = context.dummy.client.submit(JournalInsert(Entity(Map()))).join()
+    val op = context.dummy.client.insert(Entity(Map()))
+    val res = Await.result(op, timeout)
     res.foreach((entry: CanonicalEntry) => {
       context.entries += entry
       context.entityRef = entry.ref
     })
     val ref = context.entityRef
     for (x <- 1 to 5) {
```
Comment (bigs) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala]: `for (_ <- 1 to 5) {`

Comment (vyzo) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala]: ok, will underscore the unused variable

File: transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala [hunk 4]
```diff
@@ -35,26 +37,30 @@ object JournalBlockchainSpec extends io.mediachain.BaseSpec
...
     (block.index must_== 0) and
     (block.chain must beNone) and
     (block.entries must beEmpty)
   }
   
   def accumulateBlock = {
     val context = JournalBlockchainSpecContext.context
-    val res = context.dummy.client.submit(JournalInsert(Entity(Map()))).join()
+    val op = context.dummy.client.insert(Entity(Map()))
+    val res = Await.result(op, timeout)
```
Comment (bigs) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala:50]: perfect :+1:

File: transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala (hunk 3 above)
Comment (parkan) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala]: I like `(1 to 5).foreach { _ => doIt() }`

**Comment by bigs** [comment]
LGTM :smile:

File: transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala (hunk 3 above)
Comment (vyzo) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala]: i kind of prefer the for loops when it is a multiline block, as I don't have to write the fat arrow.

File: transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala (hunk 4 above)
Comment (parkan) [inline transactor/src/test/scala/io/mediachain/transactor/JournalBlockchainSpec.scala:50]: :+1:

**Pull request merged**